c_src/quicer_reg.c

```diff
@@ -16,72 +16,137 @@ limitations under the License.
-      if (IS_SAME_TERM(eprofile, ATOM_QUIC_EXECUTION_PROFILE_LOW_LATENCY))
-        {
-          RegConfig.ExecutionProfile = QUIC_EXECUTION_PROFILE_LOW_LATENCY;
-        }
-      else if (IS_SAME_TERM(eprofile,
-                            ATOM_QUIC_EXECUTION_PROFILE_TYPE_MAX_THROUGHPUT))
-        {
-          RegConfig.ExecutionProfile
-              = QUIC_EXECUTION_PROFILE_TYPE_MAX_THROUGHPUT;
-        }
-      else if (IS_SAME_TERM(eprofile,
-                            ATOM_QUIC_EXECUTION_PROFILE_TYPE_SCAVENGER))
-        {
-          RegConfig.ExecutionProfile = QUIC_EXECUTION_PROFILE_TYPE_SCAVENGER;
-        }
-      else if (IS_SAME_TERM(eprofile,
-                            ATOM_QUIC_EXECUTION_PROFILE_TYPE_REAL_TIME))
-        {
-          RegConfig.ExecutionProfile = QUIC_EXECUTION_PROFILE_TYPE_REAL_TIME;
-        }
-      else
+      return ERROR_TUPLE_2(ATOM_BADARG);
+    }
+
+  pthread_mutex_lock(&GRegLock);
+
+  if (argc == 1)
+    {
+      eprofile = argv[0];
+      if (!parse_reg_conf(eprofile, &RegConfig))
         {
+          pthread_mutex_unlock(&GRegLock);
           return ERROR_TUPLE_2(ATOM_BADARG);
         }
     }
 
   QuicerRegistrationCTX *r_ctx = init_r_ctx();
   if (!r_ctx)
     {
-      ERROR_TUPLE_2(ATOM_ERROR_NOT_ENOUGH_MEMORY);
+      pthread_mutex_unlock(&GRegLock);
+      return ERROR_TUPLE_2(ATOM_ERROR_NOT_ENOUGH_MEMORY);
     }
 
-  if (0 >= enif_get_string(
-          env, ename, r_ctx->name, UINT8_MAX + 1, ERL_NIF_LATIN1))
+  if (QUIC_FAILED(
+          status = MsQuic->RegistrationOpen(&RegConfig, &r_ctx->Registration)))
     {
-      deinit_r_ctx(r_ctx);
-      destroy_r_ctx(r_ctx);
-      ERROR_TUPLE_2(ATOM_BADARG);
+      res = ERROR_TUPLE_2(ATOM_STATUS(status));
+      goto exit;
+    }
+
+  G_r_ctx = r_ctx;
+  pthread_mutex_unlock(&GRegLock);
+
+  // nif owns the global registration
+  // thus not return to the erlang side
+  return ATOM_OK;
+
+exit:
+  destroy_r_ctx(r_ctx);
+  pthread_mutex_unlock(&GRegLock);
+  return res;
+}
+
+/*
+** For global registration only
+*/
+ERL_NIF_TERM
+deregistration(__unused_parm__ ErlNifEnv *env,
+               __unused_parm__ int argc,
+               __unused_parm__ const ERL_NIF_TERM argv[])
+{
+  int error_code = 0;
+  if (!MsQuic)
+    {
+      return ERROR_TUPLE_2(ATOM_BADARG);
     }
 
-  // Open Registration
+  pthread_mutex_lock(&GRegLock);
+  if (G_r_ctx && !G_r_ctx->is_released)
+    {
+      MsQuic->RegistrationShutdown(G_r_ctx->Registration, FALSE, error_code);
+      destroy_r_ctx(G_r_ctx);
+      G_r_ctx = NULL;
+    }
+  pthread_mutex_unlock(&GRegLock);
+  return ATOM_OK;
+}
+
+ERL_NIF_TERM
+new_registration2(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+
+  ERL_NIF_TERM ename = argv[0];
+  ERL_NIF_TERM eprofile = argv[1];
+  QUIC_REGISTRATION_CONFIG RegConfig
+      = { NULL, QUIC_EXECUTION_PROFILE_LOW_LATENCY };
+
+  QUIC_STATUS status;
+  ERL_NIF_TERM res = ATOM_OK;
+
+  if (!parse_reg_conf(eprofile, &RegConfig))
```

Comment:
another n.q.: should we assert `argc >= 1` before trying to use `eprofile`?

Reply:
new_registration2, 2 means two args.

Reply:
see also : https://github.com/emqx/quic/pull/215#discussion_r1328333119